### PR TITLE
Feature/exclude

### PR DESCRIPTION
**Pull request by PiteroS8**
## Description
Merged with treshold
Added new parameter `exclude-elements` to `screen` collector.

## Motivation and Context
We want to compare screenshots without specified elements.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](https://github.com/Cognifide/aet/blob/master/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the AET Contributor License Agreement.

### CODE REVIEW

File: documentation/src/main/wiki/ScreenCollector.md [hunk 1]
```diff
@@ -11,17 +11,17 @@ Screen Collector is responsible for collecting screenshot of the page or just pa
...
 
-**Note that you cannot maximize the window and specify the dimension at the same time. If no parameters provided, default browser size is set before taking screenshot.**
-
 ##### Parameters
 
 | Parameter | Value | Description | Mandatory |
 | --------- | ----- | ----------- | --------- |
 | `xpath` | xpath_to_element | Xpath to element(s) | optional (either xpath or css) |
 | `css` | css_selector_to_element | css selector to element(s)| optional (either xpath or css) |
+| `exclude-elements` | id or css selectors to elements separated by commas | If you provide elements in exclude-elements, they won't be checked by layout comparator, the difference won't be included for `pixelThreshold` and `percentageThreshold` and you won't see the difference on the mask. | no |
```
Comment (tkaik) [inline documentation/src/main/wiki/ScreenCollector.md]: `id or css selectors to elements separated by commas` - it's enough to say `CSS selector` - it supports comma-separated id or classes by default

File: core/jobs/src/main/java/com/cognifide/aet/job/common/collectors/screen/ScreenCollector.java [hunk 2]
```diff
@@ -97,6 +127,13 @@ public void setParameters(Map<String, String> params) throws ParametersException
         .isNotBlank(params.get(CSS_PARAM))) {
       setElementParams(params);
     }
+    if(params.containsKey(EXCLUDE_ELEMENT_PARAM)) {
+      if (StringUtils.isNotBlank(params.get(EXCLUDE_ELEMENT_PARAM))) {
+        namesOfExcludeElements = params.get("exclude-elements");
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/collectors/screen/ScreenCollector.java]: Please use constant also here.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 3]
```diff
@@ -95,27 +122,31 @@ private boolean areInputsIdentical(ArtifactsDAO artifactsDAO, ComparatorProperti
...
-        throw new ProcessingException(e.getMessage(), e);
-      } finally {
-        IOUtils.closeQuietly(mask);
+    InputStream mask = null;
+    try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+      ImageIO.write(imageComparisonResult.getResultImage(), "png", baos);
+      mask = new ByteArrayInputStream(baos.toByteArray());
+      String maskArtifactId = artifactsDAO.saveArtifact(properties, mask, CONTENT_TYPE);
+
+      if (isMaskWithoutDifference(imageComparisonResult) && !excludeFunctionIsOn) {
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: Switching those arguments will be more performant.

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/StepResult.java [hunk 4]
```diff
@@ -32,14 +32,22 @@
   private static final long serialVersionUID = 7758484589529051815L;
```
Comment (Skejven) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/StepResult.java]: After updating `Serializable` class, please update this property.

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/StepResult.java [hunk 5]
```diff
@@ -66,6 +74,10 @@ public void addError(String error) {
     return errors != null ? ImmutableList.copyOf(errors) : Collections.<String>emptyList();
   }
 
+  public Payload getPayload() {
```
[the remark this answers is not in the document]
Comment (PiteroS8) [reply api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/StepResult.java]: `getPayload()` from `StepResult` is used only in [`ComparatorMessageListenerImpl`](https://github.com/Cognifide/aet/blob/feature/exclude/core/worker/src/main/java/com/cognifide/aet/worker/listeners/ComparatorMessageListenerImpl.java#L94) without `if` statement. If you mean to use it further - [`LayoutComparator`](https://github.com/Cognifide/aet/blob/feature/exclude/core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java#L96) then `Optional<Payload>` is nice idea.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/collectors/screen/ScreenCollector.java [hunk 6]
```diff
@@ -56,12 +62,27 @@
 
   private final CollectorProperties properties;
 
+  private String namesOfExcludeElements;
```
Comment (tkaik) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/collectors/screen/ScreenCollector.java]: Consider changing the name of this field - it hold a CSS selector of elements to exclude, so maybe something like `excludeCssSelector`?

File: api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/exclude/Exclude.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.cognifide.aet.communication.api.metadata.exclude;
+
+import java.io.Serializable;
+import java.util.List;
+
+public class Exclude implements Serializable {
```
Comment (tkaik) [inline api/communication-api/src/main/java/com/cognifide/aet/communication/api/metadata/exclude/Exclude.java]: Please change the class name to something more meaningful - it should be a noun, not a verb

**Comment by Skejven** [comment]
After merging #293 please change the target branch to `master`.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java [hunk 8]
```diff
@@ -55,14 +59,27 @@
 
   private Double percentageThreshold;
 
+  private boolean excludeFunctionIsOn = false;
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/comparators/layout/LayoutComparator.java]: You don't have to initialise `boolean` with `false`, its default value.

File: core/jobs/src/main/java/com/cognifide/aet/job/common/collectors/screen/ScreenCollector.java [hunk 9]
```diff
@@ -56,12 +62,27 @@
...
+
   ScreenCollector(CollectorProperties properties, WebDriver webDriver, ArtifactsDAO artifactsDAO) {
     this.properties = properties;
     this.webDriver = webDriver;
     this.artifactsDAO = artifactsDAO;
   }
 
+  private List<ExcludedElement> getExcludeElementsFromWebElements(
+      List<WebElement> webElements) {
+    List<ExcludedElement> excludeExcludedElements = new LinkedList<>();
```
Comment (Skejven) [inline core/jobs/src/main/java/com/cognifide/aet/job/common/collectors/screen/ScreenCollector.java]: Please use `ArrayList` here to have better performance. You may estimate it's size with `webElements.size()`.

File: documentation/src/main/wiki/ScreenCollector.md [hunk 10]
```diff
@@ -11,17 +11,15 @@ Screen Collector is responsible for collecting screenshot of the page or just pa
...
 
-**Note that you cannot maximize the window and specify the dimension at the same time. If no parameters provided, default browser size is set before taking screenshot.**
-
 ##### Parameters
 
 | Parameter | Value | Description | Mandatory |
 | --------- | ----- | ----------- | --------- |
 | `xpath` | xpath_to_element | Xpath to element(s) | optional (either xpath or css) |
 | `css` | css_selector_to_element | css selector to element(s)| optional (either xpath or css) |
+| `exclude-elements` | css_selector_to_element | If you provide elements in exclude-elements, they won't be checked by layout comparator, the difference won't be included for `pixelThreshold` and `percentageThreshold` and you won't see the difference on the mask. | no |
```
Comment (Skejven) [inline documentation/src/main/wiki/ScreenCollector.md]: How about that description:
> Elements found with that selector will be ignored by layout comparator (they won't affect its results) but will be rendered on the report as captured.

**Pull request merged**